keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+to clarify the API conventions), but in the end decided to use Kind, to improve
+the user experience. That is, it's easier users to take the value from the `kind`
+field at the top of the YAML they are already using, and put it straight into
+these fields, rather than needing to do a kind-resource lookup for every user's
+interaction with the API.
+
+Secondly, as in the Ingress API, the target types must be clearly defined by
+each implementation, so there should be no opportunity for ambiguity.
+
+Lastly, this is only half of a handshake, so the From referent must also both
+support referencing the To referent, and must reference the one in the To field.
+
+With all of that in mind, we felt that the benefits in terms of specificity
+outweighed the costs in terms of user pain to use the more correct `Resource`.
+
+### API Spec
+
+```golang
+// ReferenceGrant identifies kinds of resources in other namespaces that are
+// trusted to reference the specified kinds of resources in the same namespace
+// as the policy.
+//
+// Each ReferenceGrant can be used to represent a unique trust relationship.
+// Additional ReferenceGrants can be used to add to the set of trusted
+// sources of inbound references for the namespace they are defined within.
+//
+// ReferenceGrant is a form of runtime verification allowing users to assert
+// which cross-namespace object references are permitted. Implementations that
+// support ReferenceGrant MUST NOT permit cross-namespace references which have
+// no grant, and MUST respond to the removal of a grant by revoking the access
+// that the grant allowed.
+//
+// Support: Core
+type ReferenceGrant struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ReferenceGrant.
+	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+
+	// Note that `Status` sub-resource has been excluded at the
+	// moment as it was difficult to work out the design.
+	// A `Status` sub-resource may be added in the future.
+}
+
+// +kubebuilder:object:root=true
+// ReferenceGrantList contains a list of ReferenceGrant.
+type ReferenceGrantList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ReferenceGrant `json:"items"`
+}
+
+// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
+// for Gateway API.
+type ReferenceGrantSpec struct {
+	// From describes the trusted namespaces and kinds that can reference the
+	// resources described in "To". Each entry in this list MUST be considered
+	// to be an additional place that references can be valid from, or to put
+	// this another way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	From []ReferenceGrantFrom `json:"from"`
+
+	// To describes the resources that may be referenced by the resources
```

Comment:
"describes the resources in this namespace" ?

Is "to" the right term?  We had a discussion a while back about the target of policies and, IIRC, the general trend was "subject" ?  https://groups.google.com/g/kubernetes-api-reviewers/c/ldmrXXQC4G4

Reply:
I personally think "From" and "To" provide a very clear distinction. I can't think of a great word to pair with "Subject" in the place of "From", but open to ideas.

Reply:
I agree that `subject` is more correct, but the antonym there is probably `object`, and having a ReferenceGrant with `subject` and `object` feels like it would get into... grammatical correctness discussions far more than we would like.

I agree that the language here should be clearer though.

Reply:
I've updated this, PTAL.

Reply:
I would not argue for subject/object - that's confusing, and I love to take analogies too far.

NetPol uses To/From but only for actual communications, and that still has been critiqued as perhaps "too cute".

Subject/From isn't too bad, but not as symmetric.  Subject/Referrer is correct but decidedly uncute.  Subject/Origin ?

I hold opinions from an API review POV, but I'd like sig-auth to own the decision :)

Reply:
Moved to "open questions" since I think we're going to need more discussion here.